micro_http/src/connection.rs

```diff
@@ -0,0 +1,634 @@
+
+                                self.body_bytes_to_be_read = request.headers.content_length();
+                                request.body = Some(Body::new(vec![]));
+                                self.state = ConnectionState::WaitingForBody;
+                                line_start_index += CRLF_LEN;
+                            }
+                        }
+                        // We have found the end of a header line
+                        Some(relative_line_end_index) => {
+                            let request = self.request.as_mut().unwrap();
+                            // The `line_end_index` relative to the whole buffer.
+                            let line_end_index = relative_line_end_index + line_start_index;
+
+                            // Get the line slice and parse it.
+                            let line = &self.buffer[line_start_index..line_end_index];
+                            match request.headers.parse_header_line(line) {
+                                // If a header is unsupported we ignore it.
+                                Ok(_) | Err(RequestError::UnsupportedHeader) => {}
+                                // If parsing the header invalidates the request, we propagate
+                                // the error.
+                                Err(e) => return Err(ConnectionError::ParsingError(e)),
+                            };
+
+                            // Update the `line_start_index` to where we finished parsing.
+                            line_start_index = line_end_index + CRLF_LEN;
+                        }
+                        // If we have an incomplete header line.
+                        None => {
+                            // If we have parsed 1024 bytes and still haven't found the header
+                            // line end sequence.
+                            if line_start_index == 0 && bytes_read == 1024 {
+                                // Header line is longer than 1024 bytes, so it is invalid.
+                                return Err(ConnectionError::ParsingError(
+                                    RequestError::InvalidHeader,
+                                ));
+                            }
+                            // Move the incomplete header line from the end of the buffer to
+                            // the beginning, so that we can append the rest of the line and
+                            // parse it in the next `try_read` call.
+                            self.reset_buffer(line_start_index, bytes_read);
+                            break;
+                        }
+                    }
+                }
+                ConnectionState::WaitingForBody => {
+                    // If what we have just read is not enough to complete the request and
+                    // there are more bytes pertaining to the body of the request.
+                    if self.body_bytes_to_be_read > bytes_read as i32 - line_start_index as i32 {
```

Comment:
Maybe we should have a (connection-level configurable) upper limit on `body_bytes_to_be_read` as well. Right now we only read `1024` bytes at a time, but with no checks we could end up buffering unbounded amounts of data and blowing up memory.